stackstorm-compose/docker-compose.yml

```diff
@@ -0,0 +1,202 @@
+      - private
+    volumes:
+      - ./st2.docker.conf:/etc/st2/st2.docker.conf:ro
+      - ./st2.user.conf:/etc/st2/st2.user.conf:ro
+      - stackstorm-packs-configs:/opt/stackstorm/configs:rw
+      - stackstorm-packs:/opt/stackstorm/packs:rw
+      - ${ST2_PACKS_DEV:-./packs.dev}:/opt/stackstorm/packs.dev:rw
+      - stackstorm-virtualenvs:/opt/stackstorm/virtualenvs:rw
+      - stackstorm-ssh:/home/stanley.ssh
+    dns_search: .
+  st2garbagecollector:
+    image: ${ST2_IMAGE_REPO:-stackstorm/}st2garbagecollector:${ST2_VERSION:-3.3dev}
+    depends_on: ["st2api"]
+    networks:
+      - private
+    volumes:
+      - ./st2.docker.conf:/etc/st2/st2.docker.conf:ro
+      - ./st2.user.conf:/etc/st2/st2.user.conf:ro
+    dns_search: .
+  st2notifier:
+    image: ${ST2_IMAGE_REPO:-stackstorm/}st2garbagecollector:${ST2_VERSION:-3.3dev}
+    depends_on: ["st2api"]
+    networks:
+      - private
+    volumes:
+      - ./st2.docker.conf:/etc/st2/st2.docker.conf:ro
+      - ./st2.user.conf:/etc/st2/st2.user.conf:ro
+    dns_search: .
+  st2resultstracker:
+    image: ${ST2_IMAGE_REPO:-stackstorm/}st2resultstracker:${ST2_VERSION:-3.3dev}
+    depends_on: ["st2api"]
+    networks:
+      - private
+    volumes:
+      - ./st2.docker.conf:/etc/st2/st2.docker.conf:ro
+      - ./st2.user.conf:/etc/st2/st2.user.conf:ro
+    dns_search: .
+  st2rulesengine:
+    image: ${ST2_IMAGE_REPO:-stackstorm/}st2rulesengine:${ST2_VERSION:-3.3dev}
+    depends_on: ["st2api"]
+    networks:
+      - private
+    volumes:
+      - ./st2.docker.conf:/etc/st2/st2.docker.conf:ro
+      - ./st2.user.conf:/etc/st2/st2.user.conf:ro
+    dns_search: .
+  st2sensorcontainer:
+    image: ${ST2_IMAGE_REPO:-stackstorm/}st2sensorcontainer:${ST2_VERSION:-3.3dev}
+    depends_on: ["st2api"]
+    networks:
+      - private
+    dns_search: .
+    volumes:
+      - ./st2.docker.conf:/etc/st2/st2.docker.conf:ro
+      - ./st2.user.conf:/etc/st2/st2.user.conf:ro
+      - stackstorm-virtualenvs:/opt/stackstorm/virtualenvs:ro
+      - stackstorm-packs:/opt/stackstorm/packs:ro
+      - stackstorm-packs-configs:/opt/stackstorm/configs:ro
+      - ${ST2_PACKS_DEV:-./packs.dev}:/opt/stackstorm/packs.dev:ro
+  st2timersengine:
+    image: ${ST2_IMAGE_REPO:-stackstorm/}st2timersengine:${ST2_VERSION:-3.3dev}
+    depends_on: ["st2api"]
+    networks:
+      - private
+    dns_search: .
+    volumes:
+      - ./st2.docker.conf:/etc/st2/st2.docker.conf:ro
+  st2client:
+    image: ${ST2_IMAGE_REPO:-stackstorm/}st2:${ST2_VERSION:-3.3dev}
+    networks:
+      - private
+    environment:
+      ST2_AUTH_URL: ${ST2_AUTH_URL:-http://st2auth:9100/}
+      ST2_API_URL: ${ST2_API_URL:-http://st2api:9101/}
+      ST2_STREAM_URL: ${ST2_STREAM_URL:-http://st2stream:9102/}
+    volumes:
+      - ./st2-cli.conf:/root/.st2/config
+  # external services
+  mongo:
+    image: mongo:3.4
+    networks:
+      - private
+    volumes:
+      - stackstorm-mongodb:/data/db
+    dns_search: .
+  rabbitmq:
```

Comment:
I haven't seen outright startup failure in my testing. Most of the containers depend on rabbitmq and most will throw ECONNREFUSED during cold (re)start, but once RabbitMQ initializes they all picked right up. Before I setup the "depends on", order the logs were spammed with ECONNREFUSED.

Are you seeing st2api error out and then you have to restart it by hand?  We could be seeing different results based on speed and cpu of the host. 

stackstorm-ha containers behave in a manner similiar to the suggested `restart: on-failure` option. When I do a deploy, most of the containers may end up restarting 2-8 times while waiting for rabbitmq to get its act together (and if it doesn't, they go into CrashLoopBackOff).

Reply:
I had an eye on this as well. My oberservations basically confirm what @ytjohn said.

1) The rabbitmq container restarts faster than the st2api service. Therefore we see ECONNREFUSED during the shutdown of st2api.
2) if the st2api container can not connect to the rabbitmq host for any reason (I prodoced ECONNREFUSED errors as well as EHOSTUNREACH errors) st2api recovers on it's own as soon as the connection works again

No manual intervention was used on any of the test cases described above.

Reply:
@winem @ytjohn I don't know why I'm facing another scenario unlike yours.  I'm feeling this is an improvement done in st2 3.3-dev. 

Let me write my environmental details:  
* ST2_VERSION=3.2.0 (I think thats different from yours)
* MBP 2017
* OSX 10.14.6 Mojave
* Docker for Mac 2.3.2.0 with Engine 19.03.8 and Compose 1.25.5

The interesting part is that, I had to **lower** the Docker for mac VM settings to reproduce the issue. To clarify, I made 1 CPU, 2 GB Memory and 1 GB of Swap and st2api failed on boot.  

After st2api fails on boot, it didn't retry connecting unlike yours, **because the container itself is dead**, as you can see in the below shell output:  

**docker-compose ps**
```shell
                  Name                                Command                   State                         Ports
------------------------------------------------------------------------------------------------------------------------------------
stackstorm-compose_mongo_1                 docker-entrypoint.sh mongod      Up               27017/tcp
stackstorm-compose_rabbitmq_1              docker-entrypoint.sh rabbi ...   Up               25672/tcp, 4369/tcp, 5671/tcp, 5672/tcp
stackstorm-compose_redis_1                 docker-entrypoint.sh redis ...   Up               6379/tcp
stackstorm-compose_st2actionrunner_1       /opt/stackstorm/st2/bin/st ...   Up
stackstorm-compose_st2api_1                /opt/stackstorm/st2/bin/st ...   Exit 1
stackstorm-compose_st2auth_1               /opt/stackstorm/st2/bin/st ...   Up               9100/tcp
stackstorm-compose_st2client_1             /bin/bash                        Exit 0
stackstorm-compose_st2garbagecollector_1   /opt/stackstorm/st2/bin/st ...   Up
stackstorm-compose_st2notifier_1           /opt/stackstorm/st2/bin/st ...   Up
stackstorm-compose_st2resultstracker_1     /opt/stackstorm/st2/bin/st ...   Up
stackstorm-compose_st2rulesengine_1        /opt/stackstorm/st2/bin/st ...   Up
stackstorm-compose_st2scheduler_1          /opt/stackstorm/st2/bin/st ...   Up
stackstorm-compose_st2sensorcontainer_1    /opt/stackstorm/st2/bin/st ...   Up
stackstorm-compose_st2stream_1             /opt/stackstorm/st2/bin/st ...   Up               9102/tcp
stackstorm-compose_st2timersengine_1       /opt/stackstorm/st2/bin/st ...   Up
stackstorm-compose_st2web_1                /bin/bash -c if [ ${ST2WEB ...   Up (unhealthy)   443/tcp, 127.0.0.1:8000->80/tcp
stackstorm-compose_st2workflowengine_1     /opt/stackstorm/st2/bin/st ...   Up
```

Reply:
I've tried some more, including shutting down mongo. With mongo down, st2scheduler will eventually fail after so many retries. 

I set all of the containers to `restart: on-failure`.

Reply:
@TunaYagci, I am wondering if this may be something specific to Mac or your docker version. But this is as unqualified as it gets.

I'll do a bunch of tests with the latest changes from @ytjohn later tonight. The most recent ones on Friday and Saturday evening went fine.